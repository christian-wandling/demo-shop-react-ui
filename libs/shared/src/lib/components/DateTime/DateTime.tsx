import { format, parseISO } from 'date-fns';
import { toZonedTime } from 'date-fns-tz';
import { DateTimeProps } from '../../models/dateTimeProps';

/**
 * A component that displays a formatted date and time.
 *
 * This component takes a date input and formats it according to the provided pattern
 * and timezone.
 *
 * @example
 * <DateTime
 *   dateTime={myDate}
 *   pattern="yyyy-MM-dd"
 *   timezone="America/New_York"
 * />
 */
export const DateTime = ({
  dateTime,
  className = '',
  pattern = 'MMM dd, yyyy',
  timezone = 'UTC',
  'data-testid': dataTestId = 'date-time',
}: DateTimeProps) => {
  const date = typeof dateTime === 'string' ? parseISO(dateTime) : dateTime;

  const formattedDate = format(toZonedTime(date, timezone), pattern);
  const isoString = date.toISOString();

  return (
    <time className={className} dateTime={isoString} data-testid={dataTestId}>
      {formattedDate}
    </time>
  );
};
