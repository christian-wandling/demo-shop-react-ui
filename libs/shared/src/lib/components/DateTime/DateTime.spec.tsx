import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { DateTime } from './DateTime';

describe('DateTime', () => {
  it('should render with default formatting for a Date object', () => {
    const date = new Date('2023-05-15T12:30:00Z');
    render(<DateTime dateTime={date} />);

    const element: HTMLTimeElement = screen.getByTestId('time');

    expect(element.textContent).toBe('May 15, 2023');
    expect(element.dateTime).toBe('2023-05-15T12:30:00.000Z');
  });

  it('should render with default formatting for a string date', () => {
    const dateString = '2023-05-15T12:30:00Z';
    render(<DateTime dateTime={dateString} />);

    const element: HTMLTimeElement = screen.getByTestId('time');

    expect(element.textContent).toBe('May 15, 2023');
    expect(element.dateTime).toBe('2023-05-15T12:30:00.000Z');
  });

  it('should format date according to custom pattern', () => {
    const date = new Date('2023-05-15T12:30:00Z');
    const pattern = 'yyyy-MM-dd HH:mm';
    render(<DateTime dateTime={date} pattern={pattern} />);

    const element: HTMLTimeElement = screen.getByTestId('time');

    expect(element.textContent).toBe('2023-05-15 12:30');
  });

  it('should format date according to custom timezone', () => {
    const date = new Date('2023-05-15T12:30:00Z');
    render(<DateTime dateTime={date} timezone="America/New_York" />);

    const element: HTMLTimeElement = screen.getByTestId('time');

    expect(element.textContent).toBe('May 15, 2023');
  });

  it('should format date with custom pattern and timezone', () => {
    const date = new Date('2023-05-15T12:30:00Z');
    render(<DateTime dateTime={date} pattern="yyyy-MM-dd HH:mm zzz" timezone="Asia/Tokyo" />);

    const element: HTMLTimeElement = screen.getByTestId('time');

    expect(element.textContent).toMatch(/2023-05-15 21:30/);
  });

  it('matches snapshot with default props', () => {
    const date = new Date('2023-05-15T12:30:00+00:00');
    const { baseElement } = render(<DateTime dateTime={date} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('matches snapshot with custom pattern', () => {
    const date = new Date('2023-05-15T12:30:00+00:00');
    const { container } = render(<DateTime dateTime={date} pattern="EEEE, MMMM do, yyyy" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with custom timezone', () => {
    const date = new Date('2023-05-15T12:30:00+00:00');
    const { container } = render(<DateTime dateTime={date} timezone="Europe/Paris" />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
