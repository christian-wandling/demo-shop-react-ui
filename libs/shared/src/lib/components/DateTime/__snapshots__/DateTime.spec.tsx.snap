// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateTime > matches snapshot with custom pattern 1`] = `
<time
  class=""
  data-testid="time"
  datetime="2023-05-15T12:30:00.000Z"
>
  Monday, May 15th, 2023
</time>
`;

exports[`DateTime > matches snapshot with custom timezone 1`] = `
<time
  class=""
  data-testid="time"
  datetime="2023-05-15T12:30:00.000Z"
>
  May 15, 2023
</time>
`;

exports[`DateTime > matches snapshot with default props 1`] = `
<body>
  <div>
    <time
      class=""
      data-testid="time"
      datetime="2023-05-15T12:30:00.000Z"
    >
      May 15, 2023
    </time>
  </div>
</body>
`;
